Docs/documentation 2.0 (#158)
sse-secure-systems/connaisseur: docs/basics.md

# Basics

This page explains how to configure, setup and manage Connaisseur.
In addition, the core concepts are introduced.

## Admission Control, Validators and Image policy

Connaisseur works as a [mutating admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/).
It intercepts all *CREATE* and *UPDATE* resource requests for *Pods*, *Deployments*, *ReplicationControllers*, *ReplicaSets*, *DaemonSets*, *StatefulSets*, *Jobs*, and *CronJobs* and extracts all image references for validation.
Validation relies on two core concepts: image policy and validators.
A validator is a set of configuration options required for validation like the type of signature, public key to use for verification, path to signature data, or authentication.
The image policy defines a set of rules which maps different images to those validators.
This is done via glob matching of the image name which for example allows to use different validators for different registries, repositories, images or even tags.
This is specifically useful when using public or external images from other entities like Docker's official images.

## Using Connaisseur

> Some general administratin tasks like deployment or uninstallion when using Connaisseur are described in this section.

Reviewer: ```suggestion
Some general administration tasks like deployment or uninstallation when using Connaisseur are described in this section.
```

### Requirements

Using Connaisseur requires [Git](https://git-scm.com/), a [Kubernetes](https://kubernetes.io/) cluster, [Helm](https://helm.sh/) and [Docker](https://www.docker.com/) (only for local setups) to be installed and setup.

### Get the code

> The Connaisseur source code can be cloned directly from GitHub and includes the application and helm charts in a single repository.

Reviewer: ```suggestion
The Connaisseur source code can be cloned directly from GitHub and includes the application and Helm charts in a single repository.
```

```bash
git clone https://github.com/sse-secure-systems/connaisseur.git
```

### Configure

The configuration of Connaisseur is completely done in the `helm/values.yaml`.
The upper `deployment` section offers some general Kubernetes typical configurations like image version or resources.
The configuration of `validators` and image `policy` is described in detail [below](#detailed-configuration) and stepwise in the [getting started guide](getting_started.md).
Features are described on the [respective pages](features/README.md).

Connaisseur ships with a pre-configuration that does not need any adjustments for testing.
However, validating your own images requires additional configuration.

### Deploy

Install Connaisseur by using Helm (alternatively use [Makefile](#makefile)).

```bash
helm install connaisseur helm --atomic --create-namespace --namespace connaisseur
```

This will install Connaisseur in its own namespace called `connaisseur`.
The installation itself may take some minutes, as the installation order of the Connaisseur components is critical.

> Only when the Connaisseur pods are ready and running, the Admission Webhook can be applied for intercepting requests.

Reviewer: either capitalize all the resources (`Pods`), or not (`admission webhook`)

> Error from server: admission webhook "connaisseur-svc.connaisseur.svc" denied the request (...).

kubectl run hello-world --image=docker.io/hello-world
> pod/hello-world created
```

### Delete

Just like for installation, Helm can also be used to delete Connaisseur from your cluster.

```bash
helm uninstall connaisseur -n connaisseur
```

In case uninstallation fails or problems occur during subsequent installation, you can manually remove all resources:

```bash
kubectl delete all,mutatingwebhookconfigurations,clusterroles,clusterrolebindings,configmaps,imagepolicies,secrets,serviceaccounts,customresourcedefinitions -lapp.kubernetes.io/instance=connaisseur
kubectl delete namespaces connaisseur
```

Connaisseur for example also installs a *CutstomResourceDefinition* `imagepolicies.connaisseur.policy` that validates its configuration.
In case of major releases, the configuration structure might change which causes installation to fail and you might have to delete it manually.

### Makefile

Alternatively to using Helm, you can also run the Makefile for installing, deleting and more. Here the available commands:

- `make install` -- Runs the `helm install` command.

> - `make uninstall` -- Switches the namespace to `connaisseur`, uses the `helm uninstall` command and deletes the connaisseur namespace.

Reviewer: ```suggestion
- `make uninstall` -- Switches the namespace to `connaisseur`, uses the `helm uninstall` command and deletes the `connaisseur` namespace.
```
or capitalize it

This also means there can only be one validator named `default` and for the trust roots, there can only be one called `default` within a single validator.

#### Example

```yaml
validators:
- name: default
  type: notaryv1
  host: notary.docker.io
  trust_roots:
  - name: default
    type: key
    key: |
      -----BEGIN PUBLIC KEY-----
      MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEsx28WV7BsQfnHF1kZmpdCTTLJaWe
      d0CA+JOi8H4REuBaWSZ5zPDe468WuOJ6f71E7WFg3CVEVYHuoZt2UYbN/Q==
      -----END PUBLIC KEY-----
  auth:
    username: superuser
    password: lookatmeimjumping
- name: cosigner
  type: cosign
  trust_roots:
  - name: mykey
    key: |
      -----BEGIN PUBLIC KEY-----
      MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEIFXO1w6oj0oI2Fk9SiaNJRKTiO9d
      ksm6hFczQAq+FDdw0istEdCwcHO61O/0bV+LC8jqFoomA28cT+py6FcSYw==
      -----END PUBLIC KEY-----
```

#### Static Validators

Static validators are a special type of validator that does not validate any signatures.
Depending on the `approve` value being `true` or `false`, they either allow or deny all images for which they are specified as validator.
This for example allows to implement *allowlist* or *denylist*.

> ##### Configuration options

Reviewer: Ssstatic VVVValidators vs CCCConfiguration ooooptions; same below. Consistency

| `with` | - | | Additional parameters to use for a validator. See more specifics in [validator section](validators/README.md). |
|`with.trust_root`| `default` | | The name of a trust root, which is specified within the referenced validator. If not provided, the trust root with name `default` is used if it exists. |

#### Example

```yaml
policy:
- pattern: "*:*"
- pattern: "docker.io/myrepo/*:*"
  validator: myvalidator
  with:
    trust_root: mykey
- pattern: "docker.io/myrepo/deniedimage:*"
  validator: deny
- pattern: "docker.io/myrepo/allowedimage:v*"
  validator: allow
```

### Common examples

Let's look at some useful examples for the `validators` and `policy` configuration.
These can serve as a first template beyond the pre-configuration or might just be instructive to understand validators and policies.

We assume your repository is `docker.io/myrepo` and a public key has been created.
In case this repository is private, authentication would have to be added to the respective validator for example via:

```yaml
  auth:
    secret_name: k8ssecret
```

> The Kubernetes secret would have to be created separately according to the validator documentation.

Reviewer: For the `instructive` example, I wouldn't link to `separately according to the validator documentation` but rather make it easy with username/pw or a yaml comment "contains uname+pw"
